packages/drivers/odsp-driver/src/compactSnapshotParser.ts

```diff
@@ -0,0 +1,156 @@
+ * @param node - tree node to read blob section from
+ * @param dictionary - name map, used to decode path/IDs.
+ */
+function readBlobSection(node: NodeCore, dictionary: string[]) {
+    const blobs: Map<string, ArrayBuffer> = new Map();
+    for (const [idIndex, blob] of node.iteratePairs()) {
+        assert(typeof idIndex === "number", "Blob index should be a number");
+        assert(blob instanceof BlobCore, "Blob content should be of type blob");
+        const blobId = dictionary[idIndex];
+        assert(blobId !== undefined, "blob id should be present");
+        const unit8Array = blob.buffer;
+        blobs.set(blobId,
+            unit8Array.buffer.slice(unit8Array.byteOffset, unit8Array.byteOffset + unit8Array.byteLength));
+    }
+    return blobs;
+}
+
+/**
+ * Recreates ops section of the tree.
+ * @param node - tree node to read ops section from
+ */
+function readOpsSection(node: NodeCore) {
+    const ops: ISequencedDeltaOpMessage[] = [];
+    for (let i = 0; i < node.length; ++i) {
+        ops.push(JSON.parse(node.getString(i)));
+    }
+    return ops;
+}
+
+/**
+ * Recreates snapshot tree out of tree representation.
+ * @param treeNode - tree node to de-serialize from
+ * @param dictionary - name map, used to decode path/IDs.
+ */
+function readTreeSection(treeNode: NodeCore, dictionary: string[]) {
+    const tree: ISnapshotTree = {
+        blobs: {},
+        commits: {},
+        trees: {},
+    };
+    for (const [pathIndex, child] of treeNode.iteratePairs()) {
+        assert(typeof pathIndex == "number", "Tree index should be a number");
+        const path = dictionary[pathIndex];
+        assert(path !== undefined, "Path should not be undefined");
+
+        if (child instanceof NodeCore) {
+            tree.trees[path] = readTreeSection(child, dictionary);
+        } else {
+            assert(typeof child == "number", "Should be number to look in dictionary");
+            const id = dictionary[child];
+            assert(id !== undefined, "Id is out of range");
+            tree.blobs[path] = id;
+        }
+    }
+    return tree;
+}
+
+/**
+ * Converts snapshot from binary compact representation to tree/blobs/ops.
+ * @param buffer - Compact snapshot to be parsed into tree/blobs/ops.
+ * @returns - tree, blobs and ops from the snapshot.
+ */
+export function parseCompactSnapshotResponse(buffer: ReadBuffer) {
+    const builder = TreeBuilder.load(buffer);
+    let ops: ISequencedDeltaOpMessage[] | undefined;
+
+    assert(builder.length === 1, "1 root should be there");
+    const root = builder.getNode(0);
+    assert(root.length >= 4 && root.length <= 5, "4 or 5 sections should be there");
+
+    readAndValidateHeaderSection(root.getNode(0));
+
+    const dictionary = readDictionarySection(root.getNode(2));
+
+    const tree = readTreeSection(root.getNode(1), dictionary);
+
+    const blobs = readBlobSection(root.getNode(3), dictionary);
+
+    if (root.length === 5) {
```

Comment:
I'd rather make ops sub-tree required, but empty if we do not have ops.
That will allow additions to this tree in the future (if we want to have something else here, i.e. 6th element)

Reply:
Is it a suggestion for server response? Because we just parse what we get, Currently ops are not necessary.